livekit/src/room/participant/rpc.rs

```diff
@@ -0,0 +1,110 @@
+///
+/// Build-in types are included but developers may use any string, with a max length of 256 bytes.
+#[derive(Debug, Clone)]
+pub struct RpcError {
+    pub code: u32,
+    pub message: String,
+    pub data: Option<String>,
+}
+
+impl RpcError {
+    pub const MAX_MESSAGE_BYTES: usize = 256;
+    pub const MAX_DATA_BYTES: usize = 15360; // 15 KB
+
+    /// Creates an error object with the given code and message, plus an optional data payload.
+    ///
+    /// If thrown in an RPC method handler, the error will be sent back to the caller.
+    ///
+    /// Error codes 1001-1999 are reserved for built-in errors (see RpcErrorCode for their meanings).
+    pub fn new(code: u32, message: String, data: Option<String>) -> Self {
+        Self {
+            code,
+            message: truncate_bytes(&message, Self::MAX_MESSAGE_BYTES),
+            data: data.map(|d| truncate_bytes(&d, Self::MAX_DATA_BYTES)),
+        }
+    }
+
+    pub fn from_proto(proto: RpcError_Proto) -> Self {
+        Self::new(proto.code, proto.message, Some(proto.data))
+    }
+
+    pub fn to_proto(&self) -> RpcError_Proto {
+        RpcError_Proto {
+            code: self.code,
+            message: self.message.clone(),
+            data: self.data.clone().unwrap_or_default(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum RpcErrorCode {
+    ApplicationError = 1500,
+    ConnectionTimeout = 1501,
+    ResponseTimeout = 1502,
+    RecipientDisconnected = 1503,
+    ResponsePayloadTooLarge = 1504,
+    SendFailed = 1505,
+
+    UnsupportedMethod = 1400,
+    RecipientNotFound = 1401,
+    RequestPayloadTooLarge = 1402,
+}
+
+impl RpcErrorCode {
+    pub(crate) fn message(&self) -> &'static str {
+        match self {
+            Self::ApplicationError => "Application error in method handler",
+            Self::ConnectionTimeout => "Connection timeout",
+            Self::ResponseTimeout => "Response timeout",
+            Self::RecipientDisconnected => "Recipient disconnected",
+            Self::ResponsePayloadTooLarge => "Response payload too large",
+            Self::SendFailed => "Failed to send",
+
+            Self::UnsupportedMethod => "Method not supported at destination",
+            Self::RecipientNotFound => "Recipient not found",
+            Self::RequestPayloadTooLarge => "Request payload too large",
+        }
+    }
+}
+
+impl RpcError {
+    /// Creates an error object from the code, with an auto-populated message.
+    pub(crate) fn built_in(code: RpcErrorCode, data: Option<String>) -> Self {
+        Self::new(code as u32, code.message().to_string(), data)
+    }
+}
+
+/// Maximum payload size in bytes
+pub const MAX_PAYLOAD_BYTES: usize = 15360; // 15 KB
+
+/// Calculate the byte length of a string
+pub(crate) fn byte_length(s: &str) -> usize {
+    s.as_bytes().len()
+}
+
+/// Truncate a string to a maximum number of bytes
+pub(crate) fn truncate_bytes(s: &str, max_bytes: usize) -> String {
```

Comment:
Do we need this? I feel like the error doesn't need to contains the message data

Reply:
Errors need to be able to contain data, which is useful to e.g. a validation error where you list the specific fields that failed validation. I thought it's better to automatically truncate error message data rather than throwing a different error complaining that your error was malformed...

Reply:
wdyt about publish_rpc? since we already have publish_data

Reply:
gonna have a last check-in with @davidzhao on the name of this feature before merging the initial branches and publishing the forthcoming docs. will let you know where it lands!

Reply:
I went back and forth but I think rebranding this away from "RPC" isn't worth it and I think "perform" is still a pretty good verb overall for the primary action so going to leave as-is and focus on shipping it.